# Comercio de divisas Proyecto Parcial

## Objetivos

## Objetivos específicos

### Modelo que representa el problema.

Una transacción de moneda spot es un acuerdo para comprar una cantidad de una moneda usando otra moneda.

In [37]:
import pandas as pd
df = pd.DataFrame(columns=['Peso Mexicano', 'Dólar USD', 'Euro', 'Yen'], index = ['Peso Mexicano','Dólar USD','Euro','Yen'])
df.loc['Peso Mexicano', :] = [ 1, 0.048, 0.04123,4.83]
df.loc['Dólar USD', :] = [  21.1978, 1,0.88518,102.48]
df.loc['Euro', :] = [24.2149,  1.14233, 1,117.07]
df.loc['Yen', :] = [ 0.20680,0.00976, 0.00854, 1]
df


Peso Mexicano Dólar USD     Euro     Yen
Peso Mexicano             1     0.048  0.04123    4.83
Dólar USD           21.1978         1  0.88518  102.48
Euro                24.2149   1.14233        1  117.07
Yen                  0.2068   0.00976  0.00854       1

Tabla actualizada el 9 de marzo del 2020 a las 4pm.

La función a maximizar viene dada por:

$$f1= 21.1978x_{21} + 24.2149x_{31}+ 0.2068x_{41}-x_{12}-x_{13}-x_{14}$$

Las restriccones son:

$$f1= 21.1978x_{21} + 24.2149x_{31}+ 0.2068x_{41}-x_{12}-x_{13}-x_{14}\leq 100 $$
$$f1= 21.1978x_{21} + 24.2149x_{31}+ 0.2068x_{41}-x_{12}-x_{13}-x_{14}\geq 0 $$
$$f2= 0.048x_{12}+1.1423x_{32}+0.009764x_{42}-x_{21}-x_{23}-x_{24}\geq 0 $$
$$f3= 0.04123x_{13}+0.8851x_{23}+0.00854x_{43}-x_{31}-x_{32}-x_{34}\geq 0 $$
$$f4=4.83x_{14}+102.48x_{24}+117.07x_{34}-x_{41}-x_{42}-x_{43}\geq 0 $$

Las variables involucradas son:
<ul>
    
   <li>$x_{12}$ Peso a USD</li>
    <li>$x_{13}$ Peso a Euro</li>
    <li>$x_{14}$ Peso a Yen</li>
    <li>$x_{21}$ USD a Peso</li>
    <li>$x_{23}$ USD a Euro</li>
    <li>$x_{24}$ USD a Yen</li>
    <li>$x_{31}$ Euro a Peso</li>
    <li>$x_{32}$ Euro a USD</li>
    <li>$x_{34}$ Euro a Yen</li>
    <li>$x_{41}$ Yen a Peso</li>
    <li>$x_{42}$ Yen a USD</li>
    <li>$x_{43}$ Yen a Euro</li>
</ul>

Lo que pretendemos hacer maximizando (f1) es obtener la mayor ganancia en pesos mexicanos MXN para conseguir un mejor arbitraje con la transacción de divisas. Buscamos cuál será la mejor ruta para lograr eso. 

La primer restricción es que la función que pretendemos maximizar sea menor o igual a 100 pesos mexicanos. Esa es la ganancia que estamos buscando

La siguientes restricciones son para que no haya transacciones negativas por lo tanto tiene que ser mayor o igual a cero. 

## Solución del problema de optimización.

In [9]:
import scipy.optimize as opt
import numpy as np

In [30]:
# Escriba acá las matrices como arreglos de Numpy
c = -np.array([-1,-1,-1,21.1978,0,0,24.2149,0,0,0.2068,0,0])

A = np.array([[-1,-1,-1,21.1978,0,0,24.2149,0,0,0.2068,0,0],
              [1,1,1,-21.1978,0,0,-24.2149,0,0,-0.2068,0,0],
              [-0.048,0,0,1,1,1,0,-1.1423,0,0,-0.00976,0],
              [0,-0.04123,0,0,-0.8851,0,1,1,1,0,0,-0.00854],
                [0,0,-4.83,0,0,-102.48,0,0,-117.07,1,1,1]])
b = np.array([100,0,0,0,0])

In [38]:
sol = opt.linprog(c,A,b,method="simplex")
sol

     con: array([], dtype=float64)
     fun: -99.99999999999987
 message: 'Optimization terminated successfully.'
     nit: 8
   slack: array([1.27897692e-13, 1.00000000e+02, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00])
  status: 0
 success: True
       x: array([  0.        ,   0.        ,   0.        ,   0.        ,
       426.91934136,   0.        ,   4.12968875, 373.73662029,
         0.        ,   0.        ,   0.        ,   0.        ])

## Visualización de la solución del problema

<img src="./Esquema proyecto.jpeg" style="max-width: 450px;">

Concluimos con lo siguiente:
Para alcanzar la ganancia deseada de no más de $ 100 MXN se tiene que hacer las inversiones de las las siguientes divisas de la siguiente manera. 

1. Se invertirá 373.73662029 Euros a USD 
2. Después haremos la conversión de $ 426.91934136 USD de nuevo a Euros.

3. Sacamos la diferencia de los Euros obtenidos y los que teníamos. Se invertirá la diferencia de 4.12968875 Euros a MXN con lo que se obtendrá una ganancia de no más de $ 100 MXN


